fs: allow WHATWG URL and file: URLs as paths

jasnell opened the pull request:
Updates the fs module APIs to allow ~~'file://' URL strings and~~ WHATWG URL objects using 'file:' protocol to be passed as the path.

For example:

```js
const URL = require('url').URL;
const myURL = new URL('file:///C:/path/to/file');
fs.readFile(myURL, (err, data) => {});

// or
// EDIT(addaleax): this was removed from this PR
fs.readFile('file:///C:/path/to/file', (err, data) => {});
```

On Windows, file: URLs with a hostname convert to UNC paths, while file: URLs with drive letters convert to local absolute paths:

```
file://hostname/a/b/c => \\hostname\a\b\c
file:///c:/a/b/c => c:\a\b\c
```

On all other platforms, file: URLs with a hostname are unsupported and will result in a throw:

```
file://hostname/a/b/c => throw!
file:///a/b/c => /a/b/c
```

The documentation for the fs API is intentionally not updated in this commit because the URL API is still considered experimental and is not officially documented *at this time*

Note that file: URLs are *required* by spec to always be absolute paths from the file system root.

This is a semver-major commit because it changes error handling on the fs APIs.

Refs: https://github.com/nodejs/node/issues/10703

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

fs, url

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5810/

joyeecheung commented:
I think the url label can be removed here? (Also, can you take a look at https://github.com/nodejs/github-bot/pull/115 so that the bot will stop labeling these PRs with `url`?)

jasnell commented:
heh, yeah, the bot seems to be a bit aggressive about labeling things these days

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/5811/
(previous one had an error on windows)

joyeecheung commented on lib/fs.js:
```diff
@@ -211,6 +213,7 @@ fs.access = function(path, mode, callback) {
     throw new TypeError('"callback" argument must be a function');
   }
 
+  path = getPathFromURL(path);
```
The WHATWG URL parsing could escape the null byte if it's not in the hostname ('\u0000' -> '%00', e.g. `new URL('file://hostname/a/b/c\u0000')`), thus bypassing the null check, so might worth add a test for it. Is it possible for `getPathFromURL` to put null bytes in a path without one? If not, maybe the `nullCheck` can be performed before `getPathFromURL`.

jasnell replied:
good point. it's not just about null bytes, the path could contain any number of pct-encoded characters. This should run the path through a decode before returning... will fix that

joyeecheung replied:
Yes, that can be a problem for files with non-ASCII names...might worth add a test for something like `'file://hostname/文件.txt'`.

joyeecheung replied:
Also, the hostname can be escaped by domain to ASCII too. Like `'file://你好你好/a/b'`

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+function isFileUrl(path) {
+  if (typeof path !== 'string') return false;
+  if (path.length < 7) return false;
+  if ((path[0].codePointAt(0) | 0x20) !== 102) return false; // f F
```
Why not `path.codePointAt(0)`? Or `path.startsWith('file://')`?

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+  } else {
+    return path;
+  }
+  if (url.protocol !== 'file:')
```
Maybe this can be moved into the `else if (path instanceof URL)` block, because a path that starts with `file://`(`isFileUrl `) will have the file protocol.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/5813/

jasnell commented:
sigh... lol... obvious that last run was less than successful ;-) ... will investigate later on tonight

jasnell commented:
new CI: https://ci.nodejs.org/job/node-test-pull-request/5822/

addaleax commented on foobar:
```diff
@@ -0,0 +1 @@
+foobar
```
Is checking in this file intentional?

jasnell replied:
sigh

addaleax commented on lib/internal/url.js:
```diff
@@ -932,6 +935,79 @@ function urlToOptions(url) {
+    url = new URL(path);
+  } else if (path instanceof URL) {
+    if (path.protocol !== 'file:')
+      return new Error('Only file: URLs are supported');
```
I’d wrap `file:` in backticks to make this a bit more readable

addaleax commented on lib/internal/url.js:
```diff
@@ -932,6 +935,79 @@ function urlToOptions(url) {
+  }
+}
+
+function getPathFromURLOther(url) {
```
`getPathFromURLPosix` might be a good name, that would mirror `path.win32` and `path.posix`

addaleax commented on lib/internal/url.js:
```diff
@@ -932,6 +935,79 @@ function urlToOptions(url) {
+    // Otherwise, it's a local path that requires a drive letter
+    var letter = pathname.codePointAt(1) | 0x20;
+    var sep = pathname[2];
+    if (letter < 96 || letter > 122 ||
```
`letter < 97`? And maybe add comments for `a` and `z` here fore clarity, too.

jasnell replied:
doh... silly typo

addaleax commented on lib/internal/url.js:
```diff
@@ -932,6 +935,79 @@ function urlToOptions(url) {
+function getPathFromURL(path) {
+  var url;
+  if (isFileUrl(path)) {
+    url = new URL(path);
```
Are we sure this is safe? It feels like the kind of thing that could mess up any currently existing checks that userland applications run on the pathnames… Also, technically `file:///asdf` is a perfectly valid relative path on POSIX, so that alone makes this a breaking change.
I’d feel a lot better if @nodejs/security could give this a look.

And if we make decisions based on `isFileUrl(path)`, we should probably make that function accessible from userland, too.

jasnell replied:
Yep, checking for `file://` in the string input is a breaking change. We could limit this PR to only handle `URL` instances passed in and not do checking/parsing of the string. That would make it perform a bit better also. I'm ok with pulling this part back out.

jasnell replied:
I'm going to go ahead and pull out the `file://` check and parsing so that the string input is not touched at all.

jasnell commented:
@addaleax ... updated!

addaleax commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+  var hostname = url.hostname;
+  var pathname = url.pathname;
+  if (hostname !== '') {
+    // If hostname is set, then we have a UNC path
```
… where decoding of percent escapes is handled by the OS? If yes, it might be nice to have the comment extended by that. :)

jasnell replied:
I'm not sure if we should doc the changes yet or not. I still haven't even been able to land the docs for the URL object itself yet.

And good point about the handling of pct-decoding in the hostname. We should likely be doing that here instead of relying on the OS to do it.

jasnell replied:
Actually... nevermind about the pct escapes... for hostnames, that occurs while the URL is being parsed. There is a chance, however, that the hostname would be punycode encoded so I'll need to adjust for that... although I'm not entirely sure about the current support for extended characters in UNC hostnames.

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+    }
+    pathname = pathname.slice(1);
+    for (var n = 0; n < pathname.length; n++) {
+      if (pathname[n] === '%') {
```
Maybe a `&& n + 2 < pathname.length` bound check here?

jasnell replied:
There's no need for the extra check. If `n >= pathname.length`, the `pathname[n]` will return `undefined`, which causes this check to fail.

joyeecheung replied:
But when `n >= pathname.length - 2`, `pathname.codePointAt(n + 2)` (and `pathname[n+1]` later) would be out of bound, no?

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+  }
+  var pathname = url.pathname;
+  for (var n = 0; n < pathname.length; n++) {
+    if (pathname[n] === '%') {
```
Same as above.

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+        (sep !== ':' && sep !== '|')) {
+      return new Error('File URLs must specify absolute paths');
+    }
+    pathname = pathname.slice(1);
```
The first three character is known here so maybe we can start from 3?

jasnell replied:
Checking the first three characters is to verify that the path begins with a proper windows drive letter. Once we do that, we simply strip the leading `/` off. For instance, in the file URL `file:///c:/foo`, the `pathname` component will be `/c:/foo`, so once we verify that `c:` is a proper drive letter we use slice to drop the leading `/` making it `c:/foo`.

joyeecheung replied:
Ah yes, I didn't realize the pathname is passed as the `.slice(1)` version in the `decodeURIComponent()` below.

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+    pathname = pathname.slice(1);
+    for (var n = 0; n < pathname.length; n++) {
+      if (pathname[n] === '%') {
+        var third = pathname.codePointAt(n + 2) | 0x20;
```
Is there a reason for mixing `[]` and `.codePointAt` together?

jasnell replied:
`codePointAt()` is not a completely free operation performance-wise. I'm only using it where necessary.

joyeecheung commented on lib/internal/url.js:
```diff
@@ -932,6 +935,60 @@ function urlToOptions(url) {
+        }
+      }
+    }
+    return decodeURIComponent(pathname);
```
Maybe we can decode first and then search for `/` and `\`?

jasnell replied:
We cannot. Consider the following URL: `file:///c:/%2f%5c/foo`. Here the `pathname` component is `/c:/%2f%5c/`, if we decode the `pathname` first, it would become `/c://\/`, making it impossible to distinguish the decoded `/` from the legitimate `/` path separators that already exist. The only approach that works is to scan for the %2f and %5c patterns first and *then* decode.

joyeecheung commented on test/parallel/test-fs-whatwg-url.js:
```diff
@@ -0,0 +1,70 @@
+const Buffer = require('buffer').Buffer;
+
+function pathToFileURL(p) {
+  if (!path.isAbsolute(p))
```
If we only pass the result of `path.resolve` here then is this check necessary?

jasnell replied:
Not entirely necessary but it's a good fallback check just to be safe. If carries a slight performance hit but in a test that is harmless.

jasnell commented:
@addaleax ... updated! I fixed the pct encoding for paths and the punycode decoding for the hostname in the win32 variant. Thank you very much for catching that. Should be good to go now. PTAL

jasnell commented:
I will land this tomorrow if there are no further objections and CI is good

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5887/

bnoordhuis commented on lib/internal/url.js:
```diff
@@ -932,6 +935,64 @@ function urlToOptions(url) {
+    var letter = pathname.codePointAt(1) | 0x20;
+    var sep = pathname[2];
+    if (letter < 97 || letter > 122 ||   // a..z A..Z
+        (sep !== ':' && sep !== '|')) {
```
What does a pipe character signify?

jasnell replied:
The pipe has historically been allowed on Windows as an alternative to `:`. That said, I just double checked and the parser takes care of normalizing these out so we can drop the second check

bnoordhuis commented on lib/internal/url.js:
```diff
@@ -932,6 +935,64 @@ function urlToOptions(url) {
+function getPathFromURLPosix(url) {
+  if (url.hostname !== '') {
+    return new Error(
+      `File URLs on ${os.platform()} must use hostname localhost`);
```
The error message says 'localhost' but the check is for an empty string.

jasnell replied:
Yep. For file URLs, the URL parser normalizes 'localhost' to the empty string ''. They are equivalent.

```js
> new url.URL('file://localhost/a/b/c');                                                                                                                                                                                                               1
URL {                                                                                                                                                                                                                                                  1
  href: file:///a/b/c
  protocol: file:
  pathname: /a/b/c
}
> new url.URL('file:///a/b/c');
URL {
  href: file:///a/b/c
  protocol: file:
  pathname: /a/b/c
}
```

bnoordhuis commented on test/parallel/test-fs-whatwg-url.js:
```diff
@@ -0,0 +1,70 @@
+fs.readFile(new URL('file:///c:/tmp/%00test'), common.mustCall((err) => {
+  assert(err);
+  assert.strictEqual(err.message,
+                     'Path must be a string without null bytes');
```
Depends on the URL-to-path conversion being followed by a nullCheck, which leaves room for error.  Can you update test/parallel/test-fs-null-bytes.js with URL regression tests so we have more exhaustive coverage?  You can probably do something cunning to the check() function in that file.

jasnell replied:
Done

jasnell commented on lib/internal/url.js:
```diff
@@ -932,6 +935,64 @@ function urlToOptions(url) {
+    // it has been punycode encoded. We do not need to worry
+    // about percent encoding because the URL parser will have
+    // already taken care of that for us.
+    return `//${domainToUnicode(hostname)}${pathname}`;
```
Nope... 

```js
> new url.URL('file://pip-pip/a/b/c').hostname
'pip-pip'
> url.domainToUnicode(new url.URL('file://pip-pip/a/b/c').hostname)
'pip-pip'
>
```

bnoordhuis replied:
Okay, so `url.domainToUnicode()` only acts on (sub)domains that start with xn--.  Can you edit the comment to make clear the distinction between punycode decoding and IDN?  It led me to believe that `domainToUnicode(...)` is just `punycode.decode(...)`.

jasnell commented:
@bnoordhuis ... updated to address the feedback.

fengmk2 commented on lib/fs.js:
```diff
@@ -232,6 +248,8 @@ fs.accessSync = function(path, mode) {
 };
 
 fs.exists = function(path, callback) {
+  if (handleError((path = getPathFromURL(path)), callback))
```
Should not return error to callback on exists method.

jasnell replied:
I disagree here. Note that the `nullCheck()` method check implements the same basic behavior (throwing or doing a callback with an error). Given that the `getPathFromURL()` method is not performing any actual file checks and instead throws based on syntax validation of the input.

fengmk2 replied:
nullCheck use `cb` not `callback`

jasnell replied:
ah.. right, good point :-)

fengmk2 commented on lib/fs.js:
```diff
@@ -242,6 +260,7 @@ fs.exists = function(path, callback) {
 };
 
 fs.existsSync = function(path) {
+  handleError((path = getPathFromURL(path)));
```
Same here, return false instead of throw error

jasnell commented:
@fengmk2 ... updated! :-) thank you for catching that!

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6175/

TimothyGu commented:
Are there any measurable performance difference?

jasnell commented:
The impact should be minimal due to the additional `instanceof` check.

jasnell commented:
Landed in 79400bf

jasnell commented:
Ugh. Missed a failure on windows on this. reverting in https://github.com/nodejs/node/pull/11155 and reopening

italoacasas commented:
I notice this late, I remove the commits from `v7.x-staging` already.

jasnell commented:
When this lands again let's not pull it into v7 right away. Let it sit in master for a while

mathiasbynens commented:
This change might introduce security issues in existing code.

Code such as `fs.readFile(sanitize(somethingBasedOnUserInput), cb)` where `sanitize` accounts for all cases except `file://` URLs (on the assumption that those result in exceptions anyway) is now vulnerable.

addaleax commented:
> Code such as `fs.readFile(sanitize(somethingBasedOnUserInput), cb)` where `sanitize` accounts for all cases except `file://` URLs (on the assumption that those result in exceptions anyway) is now vulnerable.

We talked about this concern in https://github.com/nodejs/node/pull/10739#discussion_r95843318 – the original PR description is no longer accurate, strings starting with `file://` are not implicitly converted to URLs.

@jasnell I’ve taken the liberty to amend the corresponding parts of the PR description.

mathiasbynens commented:
Apologies, @addaleax — I should’ve read the comments on the outdated diffs before posting. Glad to hear this has been considered and addressed.

Perhaps “Note that strings starting with `file://` are not implicitly converted to URLs.” can be added to the commit message, too.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/6218/

jasnell commented:
Ok, windows looks good. Other CI failures are unrelated. Will land again on Monday.

TimothyGu commented on lib/internal/url.js:
```diff
@@ -1112,6 +1115,66 @@ function urlToOptions(url) {
+      var third = pathname.codePointAt(n + 2) | 0x20;
+      if ((pathname[n + 1] === '2' && third === 102) || // 2f 2F /
+          (pathname[n + 1] === '5' && third === 99)) {  // 5c 5C \
+        return new Error('Path must not include encoded \\ or / characters');
```
Instead of `Error`, can `TypeError` be used? It's more specific and aligns better with other errors thrown from fs and URL.

jasnell replied:
It's not a `TypeError` tho.

joyeecheung replied:
Could go either way I think, the `URL` constructor throws a `TypeError` when the url string passed into it is invalid.

jasnell replied:
Ok, that is fine. Updated the errors to `TypeError`. Went ahead and updated all of the new Errors introduced by this to `TypeError` to maintain consistency.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/6235/

jasnell commented:
CI failures are unrelated.

jasnell commented:
Landed (again) in 9549329. 
Please do not land into 7.x until this has had a few weeks in master

italoacasas commented:
We are going to do a RC for version 7 with v8 5.5 next week, do you want to
include this in the RC?
On Mon, Feb 6, 2017 at 2:06 PM James M Snell <notifications@github.com>
wrote:

> Closed #10739 <https://github.com/nodejs/node/pull/10739>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/10739#event-950850189>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AGD2F2lhXECjHrYwF2OpCUie0M1ZMkADks5rZ29CgaJpZM4LgsK2>
> .
>
-- 
Italo A. Casas
Node.js Core Member
me@italoacasas.com | (305) 393-6424

jasnell commented:
including it in the RC may be fine

italoacasas commented:
Removing the flag since I'm going to include this in `7.6 RC`

sam-github commented:
@addaleax said:

> Should this come with docs changes?

but I didn't see any follow up comments. Why is this not documented? I just found it while browsing the fs source.

addaleax commented:
@sam-github Quoting the PR description (I think this had been added later):

> The documentation for the fs API is intentionally not updated in this commit because the URL API is still considered experimental and is not officially documented at this time

sam-github commented:
I see. We can't document that fs APIs take a url.URL object when url.URL itself is undocumented. Fair enough.

addaleax commented:
To be fair, `url.URL` has been documented since this PR landed: https://nodejs.org/api/url.html#url_the_whatwg_url_api. I’ll open a `good first contribution` issue asking for docs for this feature.
